cilium/CFP-30788-managing-hubble-metrics-cardinality.md

```diff
@@ -0,0 +1,130 @@
+
+Usually, Cluster admins would be able to control cardinality of the metrics ingested by applying custom ingestion rules on the promethues configuration. However, there are scenarios where this may not be ideal, for example:
+
+- When the cluster is using a managed prometheus offering where the prometheus configuration is managed by the cloud provider and custom ingestion rules cannot be applied.
+- When a high scale cluster is generating high cardinality of metrics putting pressure on the prometheus ingestion, storage and query performance.
+- When the cluster admin wants to dynamically control the cardinality of the metrics based on the cluster state without having to restart cilium agent or prometheus server. 
+
+Such issues present a unique challenge to cluster admins and developers. By creating a mechanism to manage the cardinality of Hubble metrics dynamically based on the cluster state would help in addressing the above challenges. Recent change for [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) has inspired this proposal.
+
+## Goals
+
+* Ability to manage the cardinality of Hubble metrics dynamically.
+* Should be able to control labels and dimensions of the metrics.
+
+
+## Non-Goals
+
+* Recommendations for Prometheus configuration at ingestion time.
+
+## Proposal
+
+### Overview
+
+We propose a new config map wrapped functionality in Hubble metrics which can be used to manage the cardinality, dimensions and targets of the metrics. Drawing from [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) which uses a standalone config map to manage exporting of flows, we propose a similar mechanism. 
+
+Existing behavior of reading cilium-config for `hubble-metrics` be preserved for backward compatibility. 
+
+A new config called `enable-dynamic-hubble-metrics` be introduced which when set to true, will enable the dynamic management of Hubble metrics. When this config is set to true, the cilium agent will read the `hubble-metrics-config` config map to get the configuration for the metrics. When this config is set to false, the cilium agent will read the `cilium-config` for `hubble-metrics` configuration.
+
+### Config Map
+
+A new config map called `hubble-metrics-config` will be introduced which will have the following fields:
+
+- `enabled`: Boolean value to enable/disable the dynamic management of Hubble metrics.
+- `metrics`: List of metrics to be managed. Each metric will have the following fields:
+  - `name`: Name of the metric.
+  - `contextoptions`: List of [context options](https://docs.cilium.io/en/latest/observability/metrics/#context-options) to be applied to the metric.
+  - `labels`: List of labels to be applied to the metric.
```

Comment:
How do you expect this to work? Just hardcoded static labels?

Reply:
This the field for the optional fields each metrics support ? https://docs.cilium.io/en/latest/observability/metrics/#id12
Are these defined in contextOptions itself today ?

Reply:
Removing, it makes sense to fold them under contextOptions.

Reply:
`labels` isn't a context option; the link your referrencing is describing the labels the metric defines; they're not configurable.

Reply:
I removed it.